luigi/contrib/lsf.py

```diff
@@ -0,0 +1,362 @@
+        create_packages_archive(packages, os.path.join(self.tmp_dir, "packages.tar"))
+
+        # Now, pass onto the class's specified init_local() method.
+        self.init_local()
+
+    def init_local(self):
+        """
+        Implement any work to setup any internal datastructure etc here.
+        You can add extra input using the requires_local/input_local methods.
+        Anything you set on the object will be pickled and available on the compute nodes.
+        """
+        pass
+
+    def run(self):
+        """
+        The procedure:
+        - Pickle the class
+        - Tarball the dependencies
+        - Construct a bsub argument that runs a generic runner function with the path to the pickled class
+        - Runner function loads the class from pickle
+        - Runner class untars the dependencies
+        - Runner function hits the button on the class's work() method
+        """
+        self._init_local()
+        self._run_job()
+
+    def work(self):
+        """
+        Subclass this for where you're doing your actual work.
+
+        Why not run(), like other tasks? Because we need run to always be
+        something that the Worker can call, and that's the real logical place to
+        do LSF scheduling.
+        So, the work will happen in work().
+        """
+        pass
+
+    def _dump(self, out_dir=''):
+        """
+        Dump instance to file.
+        """
+        self.job_file = os.path.join(out_dir, 'job-instance.pickle')
+        if self.__module__ == '__main__':
+            dump_inst = pickle.dumps(self)
+            module_name = os.path.basename(sys.argv[0]).rsplit('.', 1)[0]
+            dump_inst = dump_inst.replace('(c__main__', "(c" + module_name)
+            open(self.job_file, "w").write(dump_inst)
+
+        else:
+            pickle.dump(self, open(self.job_file, "w"))
+
+    def _run_job(self):
+        """
+        Build a bsub argument that will run lsf_runner.py on the directory we've specified.
+        """
+
+        args = []
+
+        if isinstance(self.output(), list):
+            log_output = self.output()[0].path.split("/")
+        else:
+            log_output = self.output().path.split("/")
+
+        args += ["bsub", "-q", self.queue_flag]
+        args += ["-n", str(self.n_cpu_flag)]
+        args += ["-M", str(self.memory_flag)]
+        args += ["-R", "rusage[%s]" % self.resource_flag]
+        args += ["-W", str(self.runtime_flag)]
+        if self.job_name_flag:
+            args += ["-J", str(self.job_name_flag)]
+        args += ["-o", "/".join(log_output[0:-1]) + "/job.out"]
+        args += ["-e", "/".join(log_output[0:-1]) + "/job.err"]
+        args += self.extra_bsub_args.split()
+
+        # Find where our file is
+        runner_path = lsf_runner.__file__
+        # assume source is next to compiled
+        if runner_path.endswith("pyc"):
+            runner_path = runner_path[:-3] + "py"
```

Comment:
Why are you renaming pyc to py?

If the intention is to rename `.pyc` to `.py`, why don't you just remove the `c` or something?

Reply:
Looks like this might have been dev code to ensure that the latest version of the py script was called. Python should check that the pyc timestamp is newer than the py timestamp anyway so I have removed this from the code. This could always be reintroduced if there is a problem along with a debug flag.